joomla/joomla-cms
# [5.0] Smart Search: Add debugging features

### Summary of Changes
In order to gain wider adoption from third party developers, we need to provide them with improved development tools. This PR implements a few features, which become visible when the debug mode is enabled.

#### Display indexed item
First of all we want to be able to see what has been indexed for an item. When you enable debugging, the list of entries in the index will now be linked and clicking on such an entry will give you a view which displays the common data from indexed items, the type specific items, the indexed terms and the associated taxonomies. This allows you to understand if the content was indexed the way you expected it, as well as with the weighing you expected.

#### Debug indexing of one specific item
Again, when debugging is enabled, a new dropdown button in the index view appears with "Index debugging". This button goes to a new view, where you can select a finder plugin and type in an ID to index and it will then run that plugin for that ID. It will then either return what has been prepared to be indexed by the indexer in the result object or whatever output the plugin generated. This should allow a developer to better write a new finder plugin to index their content.

### Testing Instructions
1. Install Joomla and provide some sample data. Please note that this PR requires "npm i" to be run again after it has been applied!
2. Enable debugging mode in global configuration
3. Go to Smart Search and click on an entry in the list to display the data in the index for this item
4. Click on "Index > Index debugging" and select a finder plugin from the list.
5. Type in an ID, for example "60" in case you are using the testing sample data, and click on "Index"

### Notes
1. The dropdown button seems to have a styling error when using a modal button as first entry. The spacing is wrong and I couldn't find out why.
2. The Javascript is most likely "not optimal". I'm happy for corrections.
3. I'm open for better language strings. 
4. Since 4.1 is in RC, this will have to go into 4.2. Since there is no 4.2-dev branch yet, I'm pointing this towards 4.1-dev for now.
